### ISSUE TITLE
multishard_combining_reader may be missing fragments if recreated right after prefix range tombstone

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#4733]
When `shard_reader::remote_reader` stops, it saves the key of the last emitted fragment. Then if the reader needs to be recreated, it adjusts the slice so that it will emit elements right after that fragment. However, it works with `clustering_key_prefixes`. If the last fragment was a range tombstone starting at a prefix, then recreation will create a slice which skips the whole prefix, but there may be elements after that tombstone falling into the prefix which were not emitted yet.

Reproducer:
```diff
diff --git a/mutation_reader.cc b/mutation_reader.cc
index 1d5e3cb..a78a879 100644
--- a/mutation_reader.cc
+++ b/mutation_reader.cc
@@ -1109,9 +1109,9 @@ flat_mutation_reader shard_reader::remote_reader::resume_or_create_reader() {
         _reader_created = true;
         return reader;
     }
-    if (auto reader_opt = _lifecycle_policy.try_resume(std::move(_irh))) {
-        return std::move(*reader_opt);
-    }
+    //if (auto reader_opt = _lifecycle_policy.try_resume(std::move(_irh))) {
+    //    return std::move(*reader_opt);
+    //}
     return recreate_reader();
 }
 
@@ -1203,7 +1214,7 @@ future<shard_reader::fill_buffer_result> shard_reader::remote_reader::fill_buffe
         if (pending_next_partition) {
             reader.next_partition();
         }
-
+        reader.set_max_buffer_size(1);
         return do_fill_buffer(reader, timeout).then([this, &reader, &buffer, timeout] {
             buffer = reader.detach_buffer();
             // When the reader is recreated (after having been evicted) we
diff --git a/tests/mutation_source_test.cc b/tests/mutation_source_test.cc
index 58a0802..dd5516f 100644
--- a/tests/mutation_source_test.cc
+++ b/tests/mutation_source_test.cc
@@ -1505,6 +1538,14 @@ static mutation_sets generate_mutation_sets() {
             result.unequal.emplace_back(mutations{m1, m2});
             m2.partition().apply_row_tombstone(*s1, key, tomb);
             result.equal.emplace_back(mutations{m1, m2});
+
+            // Add a row which falls under the tombstone prefix.
+            auto ck2 = clustering_key_prefix::from_deeply_exploded(*s1, {data_value(bytes("ck2_0")), data_value(bytes("ck1_1")), });
+            auto ts = new_timestamp();
+            m1.set_clustered_cell(ck2, "regular_col_2", data_value(bytes("regular_col_value")), ts, ttl);
+            result.unequal.emplace_back(mutations{m1, m2});
+            m2.set_clustered_cell(ck2, "regular_col_2", data_value(bytes("regular_col_value")), ts, ttl);
+            result.equal.emplace_back(mutations{m1, m2});
         }

```

**Comment by denesb** [comment]
Very possibly the cause of one of the spurious failures observed in #4695.

**Comment by slivne** [comment]
@tgrabiec / @denesb - which releases are suspect to this ?

**Comment by denesb** [comment]
3.1 certainly, 3.0 very likely.

**Comment by bhalevy** [comment]
Discussed with @denesb, should block 3.1.

**Comment by denesb** [comment]
The fix was merged to master, I'll prepare a backport for 3.0.
